### PR TITLE
[Dialect] Introduce `microkernel` dialect

### PR DESCRIPTION
**Pull request by huanghaixin008**
Tracking [#12](https://github.com/intel/graph-compiler/issues/12)

This PR introduces `microkernel` dialect for frontend of microkernels, representing four procedures of microkernel execution: 1. dispatch (JIT); 2. prologue (like tile config); 3. execute; 4. epilogue (like tile release); Later passes could directly operate on this dialect for microkernel-related optimization & lowering.

### CODE REVIEW

File: include/gc/Dialect/Microkernel/MicrokernelDialect.h [hunk 1]
```diff
@@ -9,7 +9,9 @@
 #ifndef GC_DIALECTS_MICROKERNELDIALECT_H
 #define GC_DIALECTS_MICROKERNELDIALECT_H
 
+#include "mlir/Bytecode/BytecodeOpInterface.h"
```
Comment (kurapov-peter) [inline include/gc/Dialect/Microkernel/MicrokernelDialect.h:12]: What's bytecode here for?

File: lib/gc/Dialect/Microkernel/MicrokernelOps.cpp [hunk 2]
```diff
@@ -8,7 +8,369 @@
...
+    blockingFactor = 2;
+  } else if (memref.getElementType().isSignedInteger(8) ||
+             memref.getElementType().isUnsignedInteger(8)) {
+    blockingFactor = 4;
+  }
+  return memref.getShape().back() == blockingFactor;
+}
+
+static bool isTypeCompatible(Type outType, Type operandAType,
+                             Type operandBType) {
```
Comment (kurapov-peter) [inline lib/gc/Dialect/Microkernel/MicrokernelOps.cpp:276]: This seems to check something more specific than type compatibility. Could you please reflect that with a name change?

File: lib/gc/Dialect/Microkernel/MicrokernelOps.cpp [hunk 3]
```diff
@@ -8,7 +8,369 @@
...
+      getElementTypeOrSelf(memrefOperands[1].getType()),
+      getElementTypeOrSelf(memrefOperands[2].getType())};
+  if (!isTypeCompatible(typeOperands[2], typeOperands[0], typeOperands[1])) {
+    return brgemmOp.emitOpError()
+           << "operands types: " << typeOperands[0] << " X " << typeOperands[1]
+           << " -> " << typeOperands[2] << " are imcompatible";
+  }
+
+  // Verify the rank of the shaped operands.
+  for (size_t idx = 0; idx < memrefOperands.size(); idx++) {
```
Comment (kurapov-peter) [inline lib/gc/Dialect/Microkernel/MicrokernelOps.cpp:329]: Why is this even a loop? There are checks for each iteration and a remainder

File: lib/gc/Dialect/Microkernel/MicrokernelOps.cpp [hunk 4]
```diff
@@ -8,7 +8,369 @@
...
+      {ITAttr(Unsigned, 8), ITAttr(Signed, 8)},
+      {ITAttr(Signed, 8), ITAttr(Unsigned, 8)},
+      {ITAttr(Unsigned, 8), ITAttr(Unsigned, 8)},
+      {ITAttr(Signed, 8), ITAttr(Signed, 8)}};
+#undef FTAttr
+#undef ITAttr
+  if (!llvm::any_of(validDataTypes,
+                    [=](std::pair<TypeAttr, TypeAttr> type_pair) {
+                      return type_pair.first == dtypes[0] ||
+                             type_pair.second == dtypes[1];
```
Comment (kurapov-peter) [inline lib/gc/Dialect/Microkernel/MicrokernelOps.cpp:189]: Shouldn't this be &&?

File: lib/gc/Dialect/Microkernel/MicrokernelOps.cpp [hunk 5]
```diff
@@ -8,7 +8,369 @@
...
+  if (failed(parseDataTypeImpl(parser, result)))
+    return failure();
+  return parser.parseOptionalAttrDict(result.attributes);
+}
+
+static LogicalResult verifyBrgemmDataTypes(ArrayAttr dtypes,
+                                           BrgemmDispatchOp op) {
+  if (dtypes.size() != 2) {
+    return op->emitOpError() << "data types size should be 2";
+  }
```
Comment (kurapov-peter) [inline lib/gc/Dialect/Microkernel/MicrokernelOps.cpp:171]: llvm code style conventions
```suggestion
  if (dtypes.size() != 2)
    return op->emitOpError() << "data types size should be 2";
```
not marking other places

File: include/gc/Dialect/Microkernel/MicrokernelOps.td [hunk 6]
```diff
@@ -10,5 +10,100 @@
...
+
+  let arguments = (ins I64:$inputs);
+
+  let assemblyFormat = [{
+    `(` $inputs `)`
+    attr-dict `:` functional-type($inputs, results)
+  }];
+}
+
+def BrgemmMemRef : AnyTypeOf<[StaticMemRefRankOf<[F32, BF16, SI32, SI8, UI8], [2, 3, 4]>, I64]>;
```
Comment (kurapov-peter) [inline include/gc/Dialect/Microkernel/MicrokernelOps.td:74]: Could you please articulate somewhere here what the exact restrictions on the input type are?

File: include/gc/Dialect/Microkernel/MicrokernelEnum.td [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+//===----------------------------------------------------------------------===//
+
+#ifndef MICROKERNEL_ENUM
+#define MICROKERNEL_ENUM
+
+include "mlir/IR/EnumAttr.td"
+include "gc/Dialect/Microkernel/MicrokernelDialect.td"
+
+def Microkernel_BrgemmFlags : I64EnumAttr<
+    "BrgemmFlags", "see: microkernel_brgemm_flags",
```
Comment (kurapov-peter) [inline include/gc/Dialect/Microkernel/MicrokernelEnum.td:16]: This either needs a better link or moving the description here. IMO the latter is preferrable.

**Review by kurapov-peter — CHANGES_REQUESTED** [review]
Good overall, requires some minor changes

File: include/gc/Dialect/Microkernel/MicrokernelDialect.h (hunk 1 above)
Comment (huanghaixin008) [reply include/gc/Dialect/Microkernel/MicrokernelDialect.h]: Redundant header removed

File: lib/gc/Dialect/Microkernel/MicrokernelOps.cpp (hunk 2 above)
Comment (huanghaixin008) [reply lib/gc/Dialect/Microkernel/MicrokernelOps.cpp]: This function checks whether the input types of Memrefs are supported, currently the following types are supported:
1. f32xf32 -> f32
2. bf16xbf16 -> f32
3. [s8, u8]x[s8, u8] -> s32
Change the function name to `isTypeSupported`

File: lib/gc/Dialect/Microkernel/MicrokernelOps.cpp (hunk 3 above)
Comment (huanghaixin008) [reply lib/gc/Dialect/Microkernel/MicrokernelOps.cpp]: Remove redundant loop and change to plain checking codes.

File: lib/gc/Dialect/Microkernel/MicrokernelOps.cpp (hunk 4 above)
Comment (huanghaixin008) [reply lib/gc/Dialect/Microkernel/MicrokernelOps.cpp]: Fixed.

File: lib/gc/Dialect/Microkernel/MicrokernelOps.cpp (hunk 5 above)
Comment (huanghaixin008) [reply lib/gc/Dialect/Microkernel/MicrokernelOps.cpp]: Checked & changed all brackets in this file per conventions.

File: include/gc/Dialect/Microkernel/MicrokernelOps.td (hunk 6 above)
Comment (huanghaixin008) [reply include/gc/Dialect/Microkernel/MicrokernelOps.td]: Changed the type name to `BrgemmMemRefOrI64` to better reflect its restrictions.
Also added comments here to explain the exact restrictions for BrgemmMemRef for each operand.

File: include/gc/Dialect/Microkernel/MicrokernelEnum.td (hunk 7 above)
Comment (huanghaixin008) [reply include/gc/Dialect/Microkernel/MicrokernelEnum.td]: Description added here.

**Review by kurapov-peter — APPROVED** [review]

**Review by ZhennanQin — APPROVED** [review]

**Pull request merged**